File: InAppChat/src/main/java/io/inappchat/inappchat/user/UserModuleImpl.java
package io.inappchat.inappchat.user;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.paging.DataSource;
import androidx.paging.PagedList;
import androidx.paging.RxPagedListBuilder;
import io.inappchat.inappchat.cache.database.entity.User;
import io.inappchat.inappchat.core.event.EventHandler;
import io.inappchat.inappchat.core.type.AvailabilityStatus;
import io.inappchat.inappchat.core.type.ChatType;
import io.inappchat.inappchat.data.DataManager;
import io.inappchat.inappchat.data.common.Result;
import io.inappchat.inappchat.module.BaseModule;
import io.inappchat.inappchat.user.mapper.UserMetaDataRecord;
import io.inappchat.inappchat.user.mapper.UserRecord;
import io.inappchat.inappchat.user.repository.UserRepository;
import io.inappchat.inappchat.user.repository.UserRepositoryImpl;
import io.reactivex.BackpressureStrategy;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import java.util.List;

/** @author meeth */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class UserModuleImpl extends BaseModule implements UserModule {

  public static UserModule newInstance(DataManager dataManager, EventHandler eventHandler) {
    return new UserModuleImpl(dataManager,eventHandler);
  }

  private final UserRepository userRepository;

  private UserModuleImpl(DataManager dataManager, EventHandler eventHandler) {
    super(dataManager);
    userRepository = UserRepositoryImpl.newInstance(dataManager,eventHandler);
  }

  private PagedList.Config config() {
    return new PagedList.Config.Builder().setEnablePlaceholders(false).setPageSize(20).build();
  }

  private <K, V> Flowable<PagedList<V>> getPagedList(
      DataSource.Factory<K, V> dataSource, PagedList.BoundaryCallback<V> boundaryCallback) {
    return new RxPagedListBuilder<>(dataSource, config())
        .setBoundaryCallback(boundaryCallback)
        .buildFlowable(BackpressureStrategy.LATEST);
  }

  @SuppressWarnings("unused")
  // @Override
  public Flowable<PagedList<User>> getPagedUsers() {
    return withTenant(
        tenantId -> {
          // Perform operation on valid tenant
          // DataSource.Factory<Integer, User> usersDataSource =
          // db().userDao().getUsersFactory(tenantId);
          return getPagedList(null, new UserBoundaryCallback(data()));
        },
        Flowable::empty);
  }

  @SuppressWarnings("unused")
  // @Override
  public Flowable<PagedList<User>> getPagedUsers(String userName) {
    return withTenant(
        tenantId -> {
          // Perform operation on valid tenant
          // DataSource.Factory<Integer, User> usersDataSource =
          //    db().userDao().getUsersWithQueryFactory(tenantId, userName);
          return getPagedList(null, new UserBoundaryCallback(data()));
        },
        Flowable::empty);
  }

  @Override
  public Flowable<List<UserRecord>> getChatUsers() {
    return userRepository.getChatUsers(getTenantId());
  }

  @Override
  public Single<List<UserRecord>> getMentionedUsers() {
    return userRepository.getMentionedUsers(getTenantId());
  }

  @Override
  public Single<List<UserRecord>> getReactionedUsers(List<String> reactionUnicodes,String msgId , String threadId, ChatType chatType) {
    return userRepository.getReactionedUsers(reactionUnicodes,msgId,threadId,chatType);
  }

  @Override
  public Single<Boolean> fetchMoreUsers() {
    return withTenant(
        tenantId -> {
          // Perform operation on valid tenant
          return Single.fromCallable(
              () -> {
                final User lastUser = userRepository.getLastUserInSync(tenantId);
                final String lastUserId = lastUser != null ? lastUser.getId() : null;
                List<User> userList = userRepository.getUsersInSync(tenantId, lastUserId);
                userRepository.saveUsersInSync(userList);
                return !userList.isEmpty();
              });
        },
        () -> Single.just(Boolean.FALSE));
  }

  @Override
  public Flowable<List<UserRecord>> getNewUsers(String addUpdateOrDelete) {
    return userRepository.getNewUsers(getTenantId(), addUpdateOrDelete);
  }


  @Override
  public Flowable<UserRecord> getUserById(@NonNull String id) {
    return withTenant(tenantId -> userRepository.getUserById(tenantId, id), Flowable::empty);
  }

  @Override
  public Single<Result> logout() {
    return userRepository.logout();
  }

  @Override
  public Single<Result> updateProfile(@NonNull String profileStatus, @NonNull String mediaPath,
      @NonNull String mediaType) {
    return userRepository.updateProfile(getTenantId(),getUserId(),profileStatus,mediaPath,mediaType);
  }

  @Override
  public Single<UserRecord> getProfile() {
    return userRepository.getProfile(getAppUserId());
  }

  @Override
  public Flowable<UserRecord> getLoggedInUser() {
    return userRepository.getLoggedInUser();
  }

  @Override
  public String getUserAvailabilityStatus() {
    return db().userDao().getUserStatus(getTenantId(), getAppUserId());
  }

  @Override
  public Single setUserAvailability(AvailabilityStatus availabilityStatus) {
    return userRepository.setUserAvailability(getTenantId(), availabilityStatus);
  }

  @Override
  public Observable<UserRecord> subscribeToUserMetaData() {
    return userRepository.subscribeToUserMetaData();
  }

  @Override
  public Single<Result> removeProfilePic() {
    return userRepository.removeProfilePic(getTenantId(),getUserId());
  }

  @Override
  public Single<Result> deactivate() {
    return userRepository.deactivate();
  }

  @Override
  public Observable<UserMetaDataRecord> metaDataOn(String appUserId) {
    return userRepository.metaDataOn(appUserId);
  }

  @Override
  public Single<Result> fetchLatestUserStatus() {
    return userRepository.fetchLatestUserStatus();
  }

  @Override
  public Observable<Result> subscribeToLogout() {
    return userRepository.subscribeToLogout();
  }
}

File: InAppChat/src/main/java/io/inappchat/inappchat/user/UserModuleImpl.kt
package io.inappchat.inappchat.user

import androidx.annotation.RestrictTo
import androidx.paging.*
import androidx.paging.rxjava3.flowable
import io.inappchat.inappchat.cache.database.entity.User
import io.inappchat.inappchat.core.event.EventHandler
import io.inappchat.inappchat.core.type.AvailabilityStatus
import io.inappchat.inappchat.core.type.ChatType
import io.inappchat.inappchat.data.DataManager
import io.inappchat.inappchat.data.common.Result
import io.inappchat.inappchat.module.BaseModule
import io.inappchat.inappchat.user.mapper.UserMetaDataRecord
import io.inappchat.inappchat.user.mapper.UserRecord
import io.inappchat.inappchat.user.repository.UserRepository
import io.inappchat.inappchat.user.repository.UserRepositoryImpl
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import java.util.concurrent.Callable

/** @author meeth
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class UserModuleImpl private constructor(dataManager: DataManager, eventHandler: EventHandler) :
    BaseModule(dataManager), UserModule {
    private val userRepository: UserRepository

    init {
        userRepository = UserRepositoryImpl.newInstance(dataManager, eventHandler)
    }

    private fun config() =
        PagingConfig(20, 20, false)

    private fun <K : Any, V : Any> getPagedList(
        factory: () -> PagingSource<K, V>
    ) =
        Pager(
            config = config(),
            pagingSourceFactory = factory
        ).flowable

    val pagedUsers
        get() = this.withTenant<Flowable<PagingData<User>>>(
            { tenantId -> getPagedList { UserPagingSource(data(), userRepository) } },
            { Flowable.empty() })

    override val chatUsers: Flowable<List<UserRecord>>
        get() =
            userRepository.getChatUsers(tenantId)

    override val mentionedUsers: Single<List<UserRecord>>
        get() = userRepository.getMentionedUsers(tenantId)

    override val loggedInUser: Flowable<UserRecord>
        get() = userRepository.getLoggedInUser()

    override val profile: Single<UserRecord>
        get() = userRepository.getProfile(appUserId!!)

    override val userAvailabilityStatus: String
        get() = db().userDao().getUserStatus(tenantId, appUserId)

    override fun getReactionedUsers(
        reactionUnicodes: List<String>,
        msgId: String,
        threadId: String,
        chatType: ChatType
    ): Single<List<UserRecord>> {
        return userRepository.getReactionedUsers(reactionUnicodes, msgId, threadId, chatType)
    }

    override fun fetchMoreUsers(): Single<Boolean> {
        return withTenant(
            { tenantId ->
                Single.fromCallable(
                    Callable {
                        val lastUser: User = userRepository.getLastUserInSync(tenantId)
                        val lastUserId = if (lastUser != null) lastUser.id else null
                        val userList: List<User> =
                            userRepository.getUsersInSync(tenantId, lastUserId)
                        userRepository.saveUsersInSync(userList)
                        !userList.isEmpty()
                    })
            },
            { Single.just(false) })
    }

    override fun getNewUsers(addUpdateOrDelete: String): Flowable<List<UserRecord>> {
        return userRepository.getNewUsers(tenantId, addUpdateOrDelete)
    }

    override fun getUserById(id: String): Flowable<UserRecord> {
        return withTenant<Flowable<UserRecord>>({ tenantId ->
            userRepository.getUserById(
                tenantId,
                id
            )
        }, { Flowable.empty() })
    }

    override fun logout(): Single<Result> {
        return userRepository.logout()
    }

    override fun updateProfile(
        profileStatus: String, mediaPath: String,
        mediaType: String
    ): Single<Result> {
        return userRepository.updateProfile(
            tenantId,
            userId!!,
            profileStatus,
            mediaPath,
            mediaType
        )
    }

    override fun setUserAvailability(availabilityStatus: AvailabilityStatus): Single<*> {
        return userRepository.setUserAvailability(tenantId, availabilityStatus)
    }

    override fun subscribeToUserMetaData(): Observable<UserRecord> {
        return userRepository.subscribeToUserMetaData()
    }

    override fun removeProfilePic(): Single<Result> {
        return userRepository.removeProfilePic(tenantId, userId!!)
    }

    override fun deactivate(): Single<Result> {
        return userRepository.deactivate()
    }

    override fun metaDataOn(appUserId: String): Observable<UserMetaDataRecord> {
        return userRepository.metaDataOn(appUserId)
    }

    override fun fetchLatestUserStatus(): Single<Result> {
        return userRepository.fetchLatestUserStatus()
    }

    override fun subscribeToLogout(): Observable<Result> {
        return userRepository.subscribeToLogout()
    }

    companion object {
        @JvmStatic
        fun newInstance(dataManager: DataManager, eventHandler: EventHandler): UserModule {
            return UserModuleImpl(dataManager, eventHandler)
        }
    }
}